Sources/PackageCollections/Storage/SQLitePackageCollectionsStorage.swift

```diff
@@ -761,47 +762,47 @@ final class SQLitePackageCollectionsStorage: PackageCollectionsStorage, Closable
 
     internal func populateTargetTrie(callback: @escaping (Result<Void, Error>) -> Void = { _ in }) {
         DispatchQueue.sharedConcurrent.async(group: nil, qos: .background, flags: .assignCurrentContext) {
-            self.targetTrieReady.memoize {
-                do {
-                    // since running on low priority thread make sure the database has not already gone away
-                    switch (try self.withStateLock { self.state }) {
-                    case .disconnected, .disconnecting:
-                        callback(.success(()))
-                        return false
-                    default:
-                        break
-                    }
+            do {
+                try self.populateTargetTrieLock.withLock { // Prevents race to calling targetTrieReady.memoize
+                    _ = try self.targetTrieReady.memoize { // memoize body is only executed once
```

Comment:
if we are adding a lock here, then there is no need to also use a ThreadSafeBox since it will be a redundant lock, or iow should be safe to change targetTrieReady to a bool with the lock on top

Reply:
True. For some reason I was under the assumption that the box uses atomics, in which case it would have made sense. OTOH, if memoize would create the value while holding the lock, we could use just the box and store the trie in there.

Reply:
> OTOH, if memoize would create the value while holding the lock, we could use just the box and store the trie in there.

yes, this was a design decision I went back and forth on when adding those boxes. in some cases the former approach is better, while in others the latter.

Reply:
in this case I think the lock approach here is appropriate, and we can simplify by making self.targetTrieReady a bool

Reply:
Yes, but the lock needs to held while reading the bool as well. Using just the box would make it hard to forget that.

Reply:
Amended https://github.com/apple/swift-package-manager/pull/3522/commits/ba5b77bb5d2e1cf5ee8d3d858f8d9361a3699452